### PR TITLE
[1966] Add Selection Description in the View DSL

### LINKED ISSUE #1966: Add Selection Description in the View DSL
A selectionDescription can be created under a node tool.
The same behaviors as the one used in the current compatibility layer is expected.

### PR DESCRIPTION
**Pull request by mcharfadi**
Bug : https://github.com/eclipse-sirius/sirius-components/issues/1966
Signed-off-by: Michaël Charfadi <michael.charfadi@obeosoft.com>

# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [ ] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: packages/view/backend/sirius-components-view-emf/src/main/java/org/eclipse/sirius/components/view/emf/ViewConverter.java [hunk 1]
```diff
@@ -77,6 +94,53 @@ public List<IRepresentationDescription> convert(List<View> views, List<EPackage>
         return result;
     }
 
+    private List<IRepresentationDescription> convertSelectionsDialogues(View view, AQLInterpreter interpreter) {
```
Comment (pcdavid) [inline packages/view/backend/sirius-components-view-emf/src/main/java/org/eclipse/sirius/components/view/emf/ViewConverter.java]: `Dialogs`, not `Dialogues`

File: packages/view/backend/sirius-components-view-emf/src/main/java/org/eclipse/sirius/components/view/emf/ViewConverter.java [hunk 2]
```diff
@@ -77,6 +94,53 @@ public List<IRepresentationDescription> convert(List<View> views, List<EPackage>
...
+        .filter(org.eclipse.sirius.components.view.SelectionDescription.class::isInstance)
+        .map(org.eclipse.sirius.components.view.SelectionDescription.class::cast)
+        .map(selectionDescription -> this.convertSelectionDialogue(selectionDescription, interpreter))
+        .toList();
+    }
+
+    private IRepresentationDescription convertSelectionDialogue(org.eclipse.sirius.components.view.SelectionDescription selectionDescription, AQLInterpreter interpreter) {
+        String selectionDescriptionId = this.objectService.getId(selectionDescription);
+        // @formatter:off
+        SelectionDescription selectionDescriptionConverted = SelectionDescription.newSelectionDescription(selectionDescriptionId)
```
Comment (pcdavid) [inline packages/view/backend/sirius-components-view-emf/src/main/java/org/eclipse/sirius/components/view/emf/ViewConverter.java]: `Dialog`, not `Dialogue`

File: packages/view/backend/sirius-components-view-emf/src/main/java/org/eclipse/sirius/components/view/emf/ViewConverter.java [hunk 3]
```diff
@@ -77,6 +94,53 @@ public List<IRepresentationDescription> convert(List<View> views, List<EPackage>
         return result;
     }
 
+    private List<IRepresentationDescription> convertSelectionsDialogues(View view, AQLInterpreter interpreter) {
+        return view.getDescriptions().stream().filter(DiagramDescription.class::isInstance)
+        .flatMap(representationDescription -> this.getAllContent(representationDescription))
+        .filter(org.eclipse.sirius.components.view.SelectionDescription.class::isInstance)
+        .map(org.eclipse.sirius.components.view.SelectionDescription.class::cast)
+        .map(selectionDescription -> this.convertSelectionDialogue(selectionDescription, interpreter))
+        .toList();
```
Comment (pcdavid) [inline packages/view/backend/sirius-components-view-emf/src/main/java/org/eclipse/sirius/components/view/emf/ViewConverter.java]: Should be indented.

**Review by pcdavid — DISMISSED** [review]

File: packages/emf/backend/sirius-components-emf/src/main/java/org/eclipse/sirius/components/emf/services/ObjectService.java [hunk 4]
```diff
@@ -50,6 +50,8 @@
 @Service
 public class ObjectService implements IObjectService {
 
+    public static final String DEFAULT_ICON_PATH = "/icons/svg/Default.svg";
```
Comment (sbegaudeau) [inline packages/emf/backend/sirius-components-emf/src/main/java/org/eclipse/sirius/components/emf/services/ObjectService.java:53]: This icon is not related to `sirius-components-core`, some of our users do not have `sirius-components-compatibility` in their final application. I don't mind `sirius-components-core` having a default image but in that case, it should be in `sirius-components-core`

**Review by sbegaudeau — CHANGES_REQUESTED** [review]

**Comment by sbegaudeau** [comment]
I did not test the review for now, I have only looked at the code quickly

File: packages/emf/backend/sirius-components-emf/src/main/java/org/eclipse/sirius/components/emf/services/ObjectService.java (hunk 4 above)
Comment (sbegaudeau) [reply packages/emf/backend/sirius-components-emf/src/main/java/org/eclipse/sirius/components/emf/services/ObjectService.java:53]: I'll move the image by myself but I'll open an issue to be fixed to take into account that a default icon is now available in ObjectService#getImagePath

**Review by sbegaudeau — COMMENTED** [review]
I've moved the icon by myself, I'll wait for the build to merge the PR

File: packages/view/backend/sirius-components-view-emf/src/main/java/org/eclipse/sirius/components/view/emf/ViewConverter.java [hunk 5]
```diff
@@ -77,13 +94,51 @@ public List<IRepresentationDescription> convert(List<View> views, List<EPackage>
         return result;
     }
 
+    private List<IRepresentationDescription> convertSelectionsDialogs(View view, AQLInterpreter interpreter) {
+        return view.getDescriptions().stream().filter(DiagramDescription.class::isInstance)
+            .flatMap(representationDescription -> this.getAllContent(representationDescription))
```
Comment (gcoutable) [inline packages/view/backend/sirius-components-view-emf/src/main/java/org/eclipse/sirius/components/view/emf/ViewConverter.java]: this::getAllContent

Comment (gcoutable) [reply packages/view/backend/sirius-components-view-emf/src/main/java/org/eclipse/sirius/components/view/emf/ViewConverter.java]: Done

**Review by gcoutable — APPROVED** [review]

**Review by sbegaudeau — APPROVED** [review]

**Pull request merged**